src/spatial/nearest_neighbor.cpp

```diff
@@ -0,0 +1,281 @@
+    }
+};
+
+
+class GraphBuildingResultSet {
+    igraph_integer_t current_vertex = 0;
+    igraph_integer_t current_added = 0;
+    igraph_real_t max_distance;
+    igraph_integer_t max_neighbors;
+    igraph_integer_t *edges;
+    igraph_real_t *dists;
+
+public:
+    using DistanceType = igraph_real_t;
+
+    GraphBuildingResultSet(const igraph_integer_t neighbors, const igraph_real_t distance) :
+        max_distance(distance),
+        max_neighbors(neighbors) {}
+
+    bool addPoint(const igraph_real_t distance, const igraph_integer_t index) {
+        igraph_integer_t i;
+
+        if (index == current_vertex) {
+            return true;
+        }
+
+        for (i = current_added; i > 0; i--) {
+            if ((dists[i-1] > distance /*|| dists[i-1] == distance && index < edges[i-1]*/ )) {
+                    if (i < max_neighbors) {
+                    dists[i] = dists[i-1];
+                    edges[i] = edges[i-1];
+                }
+            } else {
+                break;
+            }
+        }
+        if (i < max_neighbors) {
+            edges[i] = index;
+            dists[i] = distance;
+        }
+        if (current_added != max_neighbors) {
+            current_added++;
+        }
+        return true;
+    }
+
+    void init(igraph_real_t *dists_, igraph_integer_t *edges_, igraph_integer_t current_vertex_) {
+        edges = edges_;
+        dists = dists_;
+        current_added = 0;
+        current_vertex = current_vertex_;
+    }
+
+    void sort() { }
+
+    igraph_integer_t size() const {
+        return current_added;
+    }
+
+    bool full() const {
+        return current_added == max_neighbors;
+    }
+
+    bool empty() const {
+        return current_added == 0;
+    }
+
+    igraph_real_t worstDist() const {
+        if (current_added < max_neighbors || current_added == 0) {
+            return  max_distance;
+        }
+        return dists[current_added-1];
+    }
+
+};
+
+template <typename Metric, igraph_integer_t Dimension>
+static igraph_error_t neighbor_helper(
+        igraph_t *graph,
+        const igraph_matrix_t *points,
+        igraph_integer_t neighbors,
+        igraph_real_t cutoff,
+        igraph_integer_t dimension,
+        igraph_bool_t directed) {
+
+    const igraph_integer_t point_count = igraph_matrix_nrow(points);
+
+    using kdTree = nanoflann::KDTreeSingleIndexAdaptor<Metric, ig_point_adaptor, Dimension>;
+
+    ig_point_adaptor adaptor(points);
+
+    kdTree tree(dimension, adaptor, nanoflann::KDTreeSingleIndexAdaptorParams(10));
+
+    tree.buildIndex();
+
+    igraph_vector_t current_point;
+    IGRAPH_VECTOR_INIT_FINALLY(&current_point, dimension);
+
+    igraph_integer_t neighbor_count = neighbors >= 0 ? neighbors : point_count;
+
+    GraphBuildingResultSet results(neighbor_count, cutoff);
+    std::vector<igraph_integer_t> neighbor_set(neighbor_count);
+    std::vector<igraph_real_t>    distances(neighbor_count);
```

Comment:
Is there any reason these two vectors can't be made part of the result class? Conceptually, they _are_ the result.  They can still be exposed to be accessible outside the class (even just made public members).

Would it simplify the code if, instead of keeping a fixed-length vector and keeping track of the actually used number of elements in `current_added`, we allowed the vector size to actually change?  After the required amount of storage is reserved for an std::vector, there isn't a cost to shrinking it and then re-growing it again. Internally, it will use the same allocated memory.